### PR TITLE
internal/contour: websocket-routes annotation support

### LINKED ISSUE #289: Enable websocket support?
I'm trying to get websockets working but it looks like it's not enabled. When I try to upgrade to a websocket connection I get ssl handshake errors currently.

The envoy docs indicate we should be able to enable it like so:
https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route.proto.html#envoy-api-field-route-routeaction-use-websocket

Nginx handles websocket endpoints like this:
https://github.com/nginxinc/kubernetes-ingress/tree/master/examples/websocket

### PR DESCRIPTION
**Pull request by glerchundi**
Fixes #289

Signed-off-by: Gorka Lerchundi Osa <glertxundi@gmail.com>

### CODE REVIEW

**Comment by glerchundi** [comment]
PTAL @davecheney.

File: internal/contour/virtualhost.go [hunk 1]
```diff
@@ -119,7 +122,7 @@ func (v *VirtualHostCache) recomputevhost(vhost string, ingresses map[metadata]*
 
 // action computes the cluster route action, a *route.Route_route for the
 // supplied ingress and backend.
-func action(i *v1beta1.Ingress, be *v1beta1.IngressBackend) *route.Route_Route {
+func action(i *v1beta1.Ingress, be *v1beta1.IngressBackend, useWebsocket *types.BoolValue) *route.Route_Route {
```
Comment (davecheney) [inline internal/contour/virtualhost.go]: please make a wrapper here, say `websocketaction` that calls `action` then applies `ca.Route.UseWebsocket = true`. See `clusteraction` in virtualhost_test.go

File: internal/contour/virtualhost.go [hunk 2]
```diff
@@ -236,3 +241,16 @@ func virtualhost(hostname string) *route.VirtualHost {
 		Domains: []string{hostname},
 	}
 }
+
+// websocketRoutes parses the annotations map for a contour.heptio.com/websocket-routes.
+// If the value is not present, or malformed, then an empty map is returned.
+func websocketRoutes(annotations map[string]string) map[string]*types.BoolValue {
```
Comment (davecheney) [inline internal/contour/virtualhost.go]: Please add a test for this function.

File: internal/contour/virtualhost.go (hunk 1 above)
Comment (glerchundi) [reply internal/contour/virtualhost.go]: `clusteraction` is being used only in `virtualhost_test.go` and I already created a sibling for websocket tests called `clusteractionwebsocket` I can rename it to `websocketaction`:

```
// clusteractionwebsocket returns a cluster action with websocket support.
func clusteractionwebsocket(name string) *route.Route_Route {
	c := clusteraction(name)
	c.Route.UseWebsocket = &types.BoolValue{Value: true}
	return c
}
```

If I create a new func `websocketaction` the loop is going to look very awful with ifs here and there. Do you really prefer this change?

```
if wr[p.Path] {
    vv.Routes = append(vv.Routes, route.Route{
        Match:  pathToRouteMatch(p),
        Action: websocketaction(ing, &p.Backend),
    })
} else {
    vv.Routes = append(vv.Routes, route.Route{
        Match:  pathToRouteMatch(p),
        Action: action(ing, &p.Backend),
    })
}
```

```
if wr["/"] {
    r := route.Route{
        Match:  prefixmatch("/"),
        Action: websocketaction(i, i.Spec.Backend),
    }
} else {
    r := route.Route{
        Match:  prefixmatch("/"),
        Action: action(i, i.Spec.Backend, wr["/"]),
    }
}
```

```
if wr[p.Path] {
    r := route.Route{
        Match:  pathToRouteMatch(p),
        Action: websocketaction(i, &p.Backend),
    }
} else {
    r := route.Route{
        Match:  pathToRouteMatch(p),
        Action: action(i, &p.Backend),
    }
}
```

Doesn't look like a very DRY code.

File: internal/contour/virtualhost.go (hunk 2 above)
Comment (glerchundi) [reply internal/contour/virtualhost.go]: Will do.

**Comment by glerchundi** [comment]
@davecheney PTAL:
* RDS test added.
* `websocketRoutes` annotations tests added.
* `action`  in `virtualhost` kept with BoolValue until you decide how to proceed

**Comment by davecheney** [comment]
This has been merged. Thank you.

**Pull request closed without merging**